src/richie/plugins/html_sitemap/cms_plugins.py

```diff
@@ -0,0 +1,166 @@
+"""HTML Sitemap plugin for DjangoCMS."""
+from functools import reduce
+
+from django.contrib.sites.models import Site
+from django.db.models import F, Q
+from django.utils import translation
+from django.utils.translation import ugettext_lazy as _
+
+from cms.constants import PUBLISHER_STATE_PENDING
+from cms.models.pagemodel import Page
+from cms.models.pluginmodel import CMSPlugin
+from cms.plugin_base import CMSPluginBase
+from cms.plugin_pool import plugin_pool
+
+from .models import HTMLSitemapPage
+
+
+def annotate_pages(pages):
+    """A generator to annotate pages with nesting information.
+
+    This is required to build a nested list of pages from the flat mptree list of pages.
+
+    Returns:
+        tuple(page, nb_open, nb_close, is_empty):
+            - page: the current page to wrap in a <li> tag,
+            - nb_open(int): number of opening <ul> tags. Should be 1 when the nesting level
+                increases, 0 otherwise,
+            - nb_close(int): number of closing </ul> tags. Should be the gap between one page
+                and the next when less nested. When reaching the last page we should close all
+                the tags still open.
+            - is_empty(boolean): True if the current <li> tag is a leaf, False otherwise. If
+                the page is not empty, it has children and we should delay closing the <li>.
+    """
+    nb_open = nb_close = 0
+    nb_pages = len(pages)
+
+    for i in range(nb_pages):
+        page = pages[i]
+
+        if i == 0:
+            nb_open = 0
+        else:
+            nb_open = max(pages[i].level - pages[i - 1].level, 0)
+
+        is_empty = True
+        if i == nb_pages - 1:
+            nb_close = pages[i].level - 1
+        else:
+            gap = pages[i].level - pages[i + 1].level
+            nb_close = max(pages[i].level - pages[i + 1].level, 0)
+            is_empty = gap >= 0
```

Comment:
`is_empty` is disturbing for a page, what about `has_children` or `is_leaf`?

Reply:
let's go back to `is_leaf`.